PhysicsTools/PatAlgos/plugins/PATTauProducer.cc

```diff
@@ -367,23 +367,74 @@ void PATTauProducer::produce(edm::Event & iEvent, const edm::EventSetup & iSetup
         	const reco::PFCandidatePtr& icand = *it;
         	ecalEnergy += icand->ecalEnergy();
-        	hcalEnergy += icand->hcalEnergy();
+        	hcalEnergy += icand->hcalEnergy();		
+		sumEtaTimesEnergy += icand->positionAtECALEntrance().eta()*icand->energy();
+                sumEnergy += icand->energy();	 
+		const reco::Track* track = 0;
+     	        if ( icand->trackRef().isNonnull() ) track = icand->trackRef().get();
+     	        else if ( icand->muonRef().isNonnull() && icand->muonRef()->innerTrack().isNonnull()  ) track = icand->muonRef()->innerTrack().get();
+     	        else if ( icand->muonRef().isNonnull() && icand->muonRef()->globalTrack().isNonnull() ) track = icand->muonRef()->globalTrack().get();
+     	        else if ( icand->muonRef().isNonnull() && icand->muonRef()->outerTrack().isNonnull()  ) track = icand->muonRef()->outerTrack().get();
+     	        else if ( icand->gsfTrackRef().isNonnull() ) track = icand->gsfTrackRef().get();
+     	        if ( track ) {
+     	          if ( track->pt() > leadChargedCandPt ) {
+     	            leadChargedCandEtaAtEcalEntrance = icand->positionAtECALEntrance().eta();
+     	            leadChargedCandPt = track->pt();
+     	          }
+                } 		
         }
         aTauPFEssential.ecalEnergy_ = ecalEnergy;
         aTauPFEssential.hcalEnergy_ = hcalEnergy;
+	aTauPFEssential.ptLeadChargedCand_ = leadChargedCandPt;      
+        aTauPFEssential.etaAtEcalEntranceLeadChargedCand_ = leadChargedCandEtaAtEcalEntrance;
+	if (sumEnergy != 0.) {
+          aTauPFEssential.etaAtEcalEntrance_ = sumEtaTimesEnergy/sumEnergy;
+        }
+	else {
+          aTauPFEssential.etaAtEcalEntrance_ = -99.;
+	}	
         float leadingTrackNormChi2 = 0;
+	float ecalEnergyLeadChargedHadrCand = -99.;
+        float hcalEnergyLeadChargedHadrCand = -99.;
+	float emFraction = -1.;
+	float myHCALenergy = 0.;
+        float myECALenergy = 0.;	
         const reco::PFCandidatePtr& leadingPFCharged = pfTauRef->leadPFChargedHadrCand();
         if(leadingPFCharged.isNonnull()) {
+		ecalEnergyLeadChargedHadrCand = leadingPFCharged->ecalEnergy();
+                hcalEnergyLeadChargedHadrCand = leadingPFCharged->hcalEnergy(); 
         	reco::TrackRef trackRef = leadingPFCharged->trackRef();
         	if(trackRef.isNonnull()) {
-        		leadingTrackNormChi2 = trackRef->normalizedChi2();
+        		leadingTrackNormChi2 = trackRef->normalizedChi2();			
+
+			std::vector<reco::PFCandidatePtr> myPFCands;
+                        myPFCands.reserve(pfTauRef->isolationPFCands().size()+pfTauRef->signalPFCands().size()); 
+                        std::copy(pfTauRef->isolationPFCands().begin(), pfTauRef->isolationPFCands().end(),
+                        std::back_inserter(myPFCands));
+                        std::copy(pfTauRef->signalPFCands().begin(), pfTauRef->signalPFCands().end(),
+                        std::back_inserter(myPFCands));
```

Comment:
how large are these isolationPFCands and signalPFCands?
This memory churn looks like is made only to save a few lines in the for loop.
range for loop would also be better than what's given below.